File: Template/Math/fastpowmatrix.cpp
#include <algorithm>

const long long MOD = 1000000007;

void mulMatrix(long long dest[2][2], long long src[2][2])
{
	long long c[2][2] = {0,0,0,0};
	for(int i = 0; i < 2; ++i)
		for(int j = 0; j < 2; ++j)
			for(int k = 0; k < 2; ++k)
				c[i][j] = (c[i][j] + (dest[i][k]*src[k][j]))%MOD;
	std::copy(&c[0][0],&c[1][2],&dest[0][0]);
}
long long fastPowMatrix(long long exp)
{
	long long res[2][2] = {
		{1,0},
		{0,1}
	}, base[2][2] = {
		{1,1},
		{1,0}
	};

	while(exp)
	{
		if(exp&1)
			mulMatrix(res,base);
		mulMatrix(base,base);
		exp >>= 1;
	}
	return res[0][1];
}

File: Template/Python/FastPowMatrix.py
# https://www.acmicpc.net/problem/10830
import sys

MOD = 1000
n, b = map(int, sys.stdin.readline().split())
temp = [[0 for i in range(n)] for j in range(n)]

def mulMatrix(dest, src):
  for i in range(0, n):
    for j in range(0, n):
      temp[i][j] = 0
      for k in range(0, n):
        temp[i][j] += (dest[i][k] * src[k][j])
      temp[i][j] %= 1000
  
  for i in range(0, n):
    for j in range(0, n):
      dest[i][j] = temp[i][j]


base = [list(map(int, sys.stdin.readline().split())) for i in range(n)]
res = [[1 if i == j else 0 for i in range(n)] for j in range(n)]

while b:
  if b&1:
    mulMatrix(res, base)
  mulMatrix(base, base)
  b >>= 1

for i in range(0, n):
  for j in range(0, n):
    print(res[i][j]%MOD, end=' ')
  print()  